### PR TITLE
SSE: Support hysteresis threshold expression

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
Introduces a new server-side expression command that lets the user use different thresholds based on the result of the previous evaluation.
The command extends the existing Threshold command and is created when the configuration contains a single condition and two evaluators.

Configuration for the new node 
```json
{
    "refId": "C",
    "datasourceUid": "__expr__",
    "model": {
        "refId": "C",
        "type": "threshold",
        "conditions": [{
            "evaluator": {
                "params": [
                    100
                ],
                "type": "gt"
            },
            "unloadEvaluator": {
                "params": [
                    30
                ],
                "type": "lt"
            },
            "loadedDimensions": {"schema":{"name":"test","meta":{"type":"fingerprints","typeVersion":[1,0]},"fields":[{"name":"fingerprints","type":"number","typeInfo":{"frame":"uint64"}}]},"data":{"values":[[1]]}}
        }],
        "expression": "B"
    }
}
```
Where the "evaluator" element is used as the `loading` condition and the `unloadEvaluator` - as unloading condition.
NOTE: _The unloading condition is inverted - which means that when it is true, the result of evaluation is 0 (i.e. normal state)._

The loaded metrics are accepted via command definition, which is a data frame that must follow a specific format:
- frame type `fingerprints`
- frame version is `1.0`
- frame has one filed of type uint64

The data frame is converted to a hash set of fingerprints, which is used during the execution. 
If the current metric's fingerprint is found in the "loaded" metrics, its value is evaluated against the "unloading" threshold. If the unloading threshold returns true, the metric is considered as unloaded, i.e., in a normal state. Otherwise - against the "loading" threshold, i.e. the same way as the regular threshold works. 

Therefore, the example above will work the following way:
- at time T0 expression B results in two metrics `{ label:A } = 101` and `{ label:B } = 101`. The result of expression C is `{ label:A } = 1` and `{ label:B } = 1`
- at time T1, the query contains the hash of metric `{label:A}`, and  expression B results in `{ label:A } = 50` and `{ label:B } = 10`, and the result of expression C is `{ label:A } = 1` and `{ label:B } = 0`

**Why do we need this feature?**
This lets users use hysteresis-like thresholds based on the history of previous evaluations, and implement tolerance to metric fluctuations to avoid alert flapping. 

**Who is this feature for?**
Alerting users 

**Which issue(s) does this PR fix?**:

Related https://github.com/grafana/grafana/issues/6202

**Special notes for your reviewer:**
- Alerting piece will be addressed in a follow-up PR
- Front-end piece will be added in a follow-up PR
- The feature is behind the flag `recoveryThreshold`
- I chose to invert the unload threshold to match the language of the expression in UI. So, API users would not be confused. 

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/expr/hyseteresis.go
Comment (yuri-tceretian) [inline pkg/expr/hyseteresis.go:88]: Here is the logic of the new command

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [inline pkg/services/ngalert/schedule/schedule.go:599]: This is how I retrieve the current state

**Comment by kylebrandt** [comment]
I think we want to wait until we have FE support to merge this, as with ML, New Warnings (for join/union drops), Pending Error handling change, and this -- we are starting to build up an in-balance in inventory of BE code in main for SSE vs FE.

**Comment by yuri-tceretian** [comment]
@grafana/alerting-frontend, I outlined the UI component but it needs to be polished. I think we will need to add some explanations to each of the thresholds. Probably get rid of Load\Unload terms and use something. 

`"Start alerting when "IS ABOVE" [ ] and keep alerting until "IS ABOVE" []`

**Comment by gillesdemey** [comment]
Can we explore adding support for a "unload" condition (term to be determined) to the existing threshold expression instead of adding an entirely new expression?

This would help with discoverability and reduces the amount of expressions we have to support / document.

**Comment by yuri-tceretian** [comment]
Team discussed the approach, and we decided to extend the threshold expression instead of creating a new one.

File: pkg/expr/transform.go
Comment (yuri-tceretian) [reply pkg/expr/transform.go:23]: The only user for now is alerting. Alerting works in two modes: standalone and cloud. In the standalone mode, this reader will be implemented in the scheduler that will combine the state manager and rule to lazily retrieve all alerting states when\if the expression tree contains hysteresis, and do not do anything if the tree does not have such expression. I have a draft PR where I implemented it https://github.com/grafana/grafana/pull/75189/files#diff-1fe393768892c6c76a43e1ac0ea9ca69ec72d2124d6d80abbede394dd49077aeR593-R616

In the cloud case, Grafana and alerting ruler are two services. The ruler sends eval request to Grafana to evaluate. I will send the loaded metrics in the request to evaluate and reader will just provide a map parsed from the request.

Comment (yuri-tceretian) [reply pkg/expr/transform.go:23]: The alternative I considered was to create a pipeline, and then traverse the nodes, try to find the hysteresis one and assign the reader to it. The problem with this approach is mutability and leaking knowledge about the server-side expression internals outside of the domain.

Comment (kylebrandt) [reply pkg/expr/transform.go:23]: > In the cloud case, Grafana and alerting ruler are two services. The ruler sends eval request to Grafana to evaluate. I will send the loaded metrics in the request to evaluate and reader will just provide a map parsed from the request.

I think:
 - Make the reader more like POST data, so standalone mode is like cloud mode from the SSE service perspective
 - Move this post-like data from the request to a query
 - Make that property data.Frames

I imagine this should result in extending the capability of a node to take input that comes from somewhere other than an a datasource or another node in a way that can be generally useful. This also keeps the stateless nature of the service intact.

**Review by kylebrandt — APPROVED** [review]

**Comment by dmarav** [comment]
Hi @yuri-tceretian ! I am trying to test this functionality in Grafana 10.2.2. I have enabled the feature toggle via configuration, and created an Alert Rule using the API with following the example in your first comment in this PR, but the alert is still resolved when the query value is lower than the executor (normal behaviour). 

Here's my config:

 ```
     {
            "refId": "B",
            "queryType": "",
            "relativeTimeRange": {
                "from": 60,
                "to": 0
            },
            "datasourceUid": "__expr__",
            "model": {
                "conditions": [
                    {
                        "evaluator": {
                            "params": [
                                120
                            ],
                            "type": "gt"
                        },
                        "unloadEvaluator": {
                            "params": [
                                1
                            ],
                            "type": "lt"
                        }
                    }
                ],
                "expression": "A",
                "intervalMs": 1000,
                "maxDataPoints": 43200,
                "refId": "B",
                "type": "threshold"
            }
        }
    ],
```

`refId` A is a prometheus query.

I saw you [PR](https://github.com/grafana/grafana/pull/75189) for the UI is waiting to be merged, and another related [PR](https://github.com/grafana/grafana/pull/79032) which seems to be related to the UI as well.

Am I missing something in the configuration or will the feature not be 100% functional until the UI is available?

Thanks in advance,
Diego

**Comment by yuri-tceretian** [comment]
@dmarav thanks for the interest in the feature! This PR introduces support for recovery threshold (aka hysteresis) only in server-side expressions. It misses the most important part - the feedback input, which provides information about the results of the previous evaluation. The missing part is in https://github.com/grafana/grafana/pull/75189. Once it is merged you can test the new expression.

**Comment by dmarav** [comment]
Thanks for the update @yuri-tceretian! I will wait until the missing part is merged and test afterwards. 

Do you have an estimate of when it will be made available? I understand it doesn't depend entirely on you but also on the Grafana release schedule, I just want to know for planning my work.

**Comment by dmarav** [comment]
Also, @yuri-tceretian please reach out to me if I can help you in any way.

**Pull request merged**
